File: Java/easy/palindromeLinkedList.java
import java.util.*;

class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

class Solution {
    private static ListNode reverse(ListNode node){
        ListNode prev = null;

        while(node != null){
            ListNode post = node.next;
            node.next = prev;
            prev = node;
            node = post;
        }

        return prev;
    }

    private static ListNode findTailOfFirst(ListNode node){
        ListNode slow = node, fast = node;

        while(fast != null && fast.next != null && fast.next.next != null){
            slow = slow.next;
            fast = fast.next.next;
        }

        return slow;
    }

    public static boolean isPalindrome(ListNode head) {
        ListNode firstHalfTail = findTailOfFirst(head);
        ListNode secondHead = reverse(firstHalfTail.next);
        
        while(head != null && secondHead != null){
            if(head.val != secondHead.val)  return false;
            head = head.next;
            secondHead = secondHead.next;
        }
        return true;
    }

    public static void printList(ListNode start){
        while(start != null){
            System.out.print(" " + start.val);
            start = start.next;
        }
    }

    // Head =  1 2 3 1 Not Palindromic    
    // Head =  1 2 2 1 Palindromic        
    // Head =  1 2 3 2 1       Palindromic

    public static void main(String[] args){
        ListNode test1 = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(1, null))));
        ListNode test2 = new ListNode(1, new ListNode(2, new ListNode(2, new ListNode(1, null))));
        ListNode test3 = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(2, new ListNode(1, null)))));
        ListNode[] testSet = new ListNode[]{test1, test2, test3};

        for(ListNode test : testSet){
            System.out.print("Head = ");
            printList(test);
            System.out.println("\t" + (isPalindrome(test)?"":"Not ") + "Palindromic");
        }
    }
}

File: CPP/easy/palindromeLinkedList.cpp
#include <vector>
#include <iostream>

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};


ListNode* reverse(ListNode* head){
    if(!head->next) return head;
    ListNode* prev = nullptr;
    ListNode* node = head;

    while(node){
        ListNode* post = node->next;
        node->next = prev;
        prev = node;
        node = post;
    }
    return prev;
}

ListNode* findFirstTail(ListNode* node){
    ListNode* slow = node;
    ListNode* fast = node;
    // fast pointer is always twice faster than the slow pointer
    // hence, this loop can find the half point of the list
    while(fast && fast->next && fast->next->next){
        fast = fast->next->next;
        slow = slow->next;
    }
    return slow;
}

bool isPalindrome(ListNode* head) {
    ListNode* firstTail = findFirstTail(head);
    // firstTail's next will become the head of the second half
    ListNode* secondHead = reverse(firstTail->next? firstTail->next : firstTail);

    // no need to care if we checked the firstTail or not
    // if we have odd numbers of node   =>  firstTail is alone and it is palindromic
    // if we have even numbers of node  =>  firstTaill will be checked when head reached the end
    while(head && secondHead){
        if(head->val != secondHead->val)    return false;
        secondHead = secondHead->next;
        head = head->next;
    }
    
    return true;
}
void printList(ListNode* start){
    while(start){
        std::cout << " " << start->val;
        start = start->next;
    }
}

// Head = 1 2 3 1  Not Palindromic
// Head = 1 2 2 1  Palindromic
// Head = 1 2 3 2 1        Palindromic

int main(){
    ListNode* test1 = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(1, nullptr))));
    ListNode* test2 = new ListNode(1, new ListNode(2, new ListNode(2, new ListNode(1, nullptr))));
    ListNode* test3 = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(2, new ListNode(1, nullptr)))));

    std::vector<ListNode*> testSet = {test1, test2, test3};
    for(ListNode* test : testSet){
        std::cout << "Head =";
        printList(test);
        std::cout << "\t" << (isPalindrome(test)? "":"Not ") << "Palindromic" << std::endl;
    }
    return 0;
}